midend/copyStructures.cpp

```diff
@@ -62,54 +62,109 @@ const IR::Node* RemoveAliases::postorder(IR::P4Control* control) {
-                        "%1%: functions or methods returning structures "
-                        "are not supported on this target",
-                        statement->right);
-            }
-            return statement;
-        }
 
-        auto* retval = new IR::IndexedVector<IR::StatOrDecl>();
-        const auto* strct = ltype->to<IR::Type_StructLike>();
-        if (const auto* list = statement->right->to<IR::ListExpression>()) {
-            unsigned index = 0;
-            for (const auto* f : strct->fields) {
-                const auto* right = list->components.at(index);
-                const auto* left = new IR::Member(statement->left, f->name);
-                retval->push_back(new IR::AssignmentStatement(statement->srcInfo, left, right));
-                index++;
-            }
-        } else if (const auto* si = statement->right->to<IR::StructExpression>()) {
-            for (const auto* f : strct->fields) {
-                const auto* right = si->components.getDeclaration<IR::NamedExpression>(f->name);
-                const auto* left = new IR::Member(statement->left, f->name);
-                retval->push_back(
-                    new IR::AssignmentStatement(statement->srcInfo, left, right->expression));
-            }
-        } else {
-            if (ltype->is<IR::Type_Header>()) {
-                // Leave headers as they are -- copy_header will also copy the valid bit
-                return statement;
-            }
-
-            BUG_CHECK(
-                statement->right->is<IR::PathExpression>() || statement->right->is<IR::Member>() ||
-                    statement->right->is<IR::ArrayIndex>(),
-                "%1%: Unexpected operation when eliminating struct copying", statement->right);
-            for (const auto* f : strct->fields) {
-                const auto* right = new IR::Member(statement->right, f->name);
-                const auto* left = new IR::Member(statement->left, f->name);
-                retval->push_back(new IR::AssignmentStatement(statement->srcInfo, left, right));
-            }
+    // If the left type is not a struct like or a header stack, return.
+    if (!(ltype->is<IR::Type_StructLike>() || ltype->is<IR::Type_Stack>())) {
+        return statement;
+    }
+
+    // Do not copy structures for method calls.
+    if (statement->right->is<IR::MethodCallExpression>()) {
+        if (errorOnMethodCall) {
+            ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                    "%1%: functions or methods returning structures "
+                    "are not supported on this target",
+                    statement->right);
         }
-        return new IR::BlockStatement(statement->srcInfo, *retval);
+        return statement;
     }
 
-    return statement;
+    const auto& srcInfo = statement->srcInfo;
+    auto* retval = new IR::IndexedVector<IR::StatOrDecl>();
+    if (const auto* list = statement->right->to<IR::ListExpression>()) {
```

Comment:
I think this cannot happen anymore, these should have been replaced with StructExpressions.
This is a leftover from when we didn't have StructExpressions.

Reply:
So where in the compiler do ListExpressions still exist? They are useful in that StructExpressions require NamedExpressions, a specific type of Expression.

Reply:
They may still exist before the type checker replaces them with struct expressions. They show up as initializers for various data structures.